### PR TITLE
B:18830 - Add Office Account Request Form

### PR DESCRIPTION
**Pull request by TevinAdams**
## Summary

This MR introduces functionality that will allow a user to request an office account by filling out a form.

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. Access the Office Sign In Page
2. Select the Request Account tab in the top right corner of the navigation bar
3. Fill out the required information on the form then select Request Account Button
4. Check database table office_users. You should see the record with a REQUESTED status. You can also check the network tab for a 200.

### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/adrs/Browser-Support/#minimum-browser-requirements) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.
- [ ] This change meets the standards for [Section 508 compliance](https://www.ssa.gov/accessibility/andi/help/install.html).

### CODE REVIEW

**Comment by danieljordan-caci** [comment]
You have a TON of commits that need to be looked at. A lot of these should already be in `integrationTesting` - do you happen to have some `main` commits in here?

File: src/utils/validation.js [hunk 1]
```diff
@@ -156,3 +156,66 @@ export const oktaInfoSchema = Yup.object().shape({
...
+    .min(10, otherUniqueIdErrorMsg)
+    .max(10, otherUniqueIdErrorMsg)
+    .matches(/^[0-9]*$/, numericOnlyErrorMsg)
+    .nullable(),
+  officeAccountRequestTelephone: phoneSchema.required('Required'),
+  officeAccountRequestEmail: emailSchema.required('Required'),
+  officeAccountTransportationOffice: Yup.object().required('Required'),
+  transportationOrderingOfficerCheckBox: Yup.bool().test(
+    'roleRequestedRequired',
+    'Please select a role to request.',
```
Comment (danieljordan-caci) [inline src/utils/validation.js:198]: Can you change `Please select a role to request` to something like `You must select at least one role`?

**Comment by TevinAdams** [comment]
> Looking at the blurb you have on the page - have you had this reviewed by Danny? Might want to make sure he's cool with what's written. Ask me how I know haha
> 
> ```
> To request an office acount, please fill out the form below with all required fields. \
> Once submitted, a MilMove admin will review your account. 
> If approved, you will be notified and allowed to log in.
> 
> NOTE: When filling out your EDIPI or unique identifier, you MUST provide one or the other. 
> If using CAC, fill out the EDIPI field. If using PIV or ECA certs, please fill out the unique identifier.
> ```

@danieljordan-caci  All texts came from the attachment in the story from the text labels to the blurb up top.

**Comment by danieljordan-caci** [comment]
> > Looking at the blurb you have on the page - have you had this reviewed by Danny? Might want to make sure he's cool with what's written. Ask me how I know haha
> > ```
> > To request an office acount, please fill out the form below with all required fields. \
> > Once submitted, a MilMove admin will review your account. 
> > If approved, you will be notified and allowed to log in.
> > 
> > NOTE: When filling out your EDIPI or unique identifier, you MUST provide one or the other. 
> > If using CAC, fill out the EDIPI field. If using PIV or ECA certs, please fill out the unique identifier.
> > ```
> 
> @danieljordan-caci All texts came from the attachment in the story from the text labels to the blurb up top.

Yeah I was the one that created that mockup but I have found that Danny doesn't like the word EDIPI so just saying to double check with him on it since we are actually pushing the code and not just doing a mockup.

File: src/components/form/OfficeAccountRequestFields/index.jsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            label="Middle Initial"
+            name={middleInitialFieldName}
+            id="officeAccountRequestMiddleInitial"
+            optional
+          />
+          <TextField label="Last Name" name={lastNameFieldName} id="officeAccountRequestLastName" />
+          <TextField label="Email" name={emailField} id="officeAccountRequestEmail" />
+          <TextField label="Telephone" name={telephoneFieldName} id="officeAccountRequestTelephone" />
+          <TextField
+            label="DoD ID number | EDIPI"
```
[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/components/form/OfficeAccountRequestFields/index.jsx:34]: @danieljordan-caci  That is what the attachment in the agility ticket had.

**Comment by TevinAdams** [comment]
> > > Looking at the blurb you have on the page - have you had this reviewed by Danny? Might want to make sure he's cool with what's written. Ask me how I know haha
> > > ```
> > > To request an office acount, please fill out the form below with all required fields. \
> > > Once submitted, a MilMove admin will review your account. 
> > > If approved, you will be notified and allowed to log in.
> > > 
> > > NOTE: When filling out your EDIPI or unique identifier, you MUST provide one or the other. 
> > > If using CAC, fill out the EDIPI field. If using PIV or ECA certs, please fill out the unique identifier.
> > > ```
> > 
> > 
> > @danieljordan-caci All texts came from the attachment in the story from the text labels to the blurb up top.
> 
> Yeah I was the one that created that mockup but I have found that Danny doesn't like the word EDIPI so just saying to double check with him on it since we are actually pushing the code and not just doing a mockup.

@danieljordan-caci Okay cool Ill change it then. I didn tknow you made it but thats an easy change.

File: src/components/form/OfficeAccountRequestFields/index.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/components/form/OfficeAccountRequestFields/index.jsx:34]: @danieljordan-caci  Yup I can make that change.

File: src/components/Office/RequestAccountForm/RequestAccountForm.jsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+              <h2>Request Office Account</h2>
+              <p>
+                To request an office acount, please fill out the form below with all required fields. Once submitted, a
+                MilMove admin will review your account. If approved, you will be notified and allowed to log in.
+              </p>
+              <p>
+                <b>NOTE:</b> When filling out your EDIPI or unique identifier, you <b>MUST</b> provide one or the other.
+                If using CAC, fill out the EDIPI field. If using PIV or ECA certs, please fill out the unique
+                identifier.
+              </p>
```
[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/components/Office/RequestAccountForm/RequestAccountForm.jsx:39]: @danieljordan-caci  Done

File: src/components/MilMoveHeader/LoggedOutUserInfo.jsx [hunk 4]
```diff
@@ -4,17 +4,36 @@ import { Button } from '@trussworks/react-uswds';
...
-      <ul className="usa-nav__primary">
-        <li className="usa-nav__primary-item">
-          <Button unstyled aria-label="Sign In" onClick={handleLogin} data-testid="signin" type="button">
-            Sign in
-          </Button>
-        </li>
-      </ul>
-    </div>
+    <>
+      {app === 'office' && (
```
[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/components/MilMoveHeader/LoggedOutUserInfo.jsx:10]: @danieljordan-caci  Done

File: src/utils/validation.js [hunk 5]
```diff
@@ -156,3 +156,66 @@ export const oktaInfoSchema = Yup.object().shape({
...
+  officeAccountRequestEdipi: Yup.string()
+    .min(10, edipiMaxErrorMsg)
+    .max(10, edipiMaxErrorMsg)
+    .matches(/^[0-9]*$/, numericOnlyErrorMsg)
+    .nullable(),
+  officeAccountRequestOtherUniqueId: Yup.string()
+    .min(10, otherUniqueIdErrorMsg)
+    .max(10, otherUniqueIdErrorMsg)
+    .matches(/^[0-9]*$/, numericOnlyErrorMsg)
+    .nullable(),
```
[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/utils/validation.js:192]: @danieljordan-caci  Done

[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/utils/validation.js:192]: @danieljordan-caci  should otherUniqueId be required if there is no DODID# filled out?

[the remark this answers is not in the document]
Comment (danieljordan-caci) [reply src/utils/validation.js:192]: Can look at it that way, yeah, but also need to also consider it the other way. If `otherUniqueID` is not provided then `DODID#` is required

File: src/components/form/OfficeAccountRequestFields/index.jsx [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    <Fieldset>
+      {render(
+        <>
+          <TextField label="First Name" name={firstNameFieldName} id="officeAccountRequestFirstName" />
+          <TextField
+            label="Middle Initial"
+            name={middleInitialFieldName}
+            id="officeAccountRequestMiddleInitial"
+            optional
+          />
```
[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/components/form/OfficeAccountRequestFields/index.jsx:29]: @danieljordan-caci  Done

File: src/components/form/OfficeAccountRequestFields/index.jsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            inputMode="numeric"
+            data-testid="officeAccountRequestEdipi"
+          />
+          <TextField
+            label="Other unique identifier"
+            labelHint="If using PIV"
+            name={otherUniqueIdName}
+            id="officeAccountRequestOtherUniqueId"
+            maxLength="10"
+            inputMode="numeric"
```
[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/components/form/OfficeAccountRequestFields/index.jsx:48]: @danieljordan-caci  Done

File: src/utils/validation.js (hunk 1 above)
Comment (TevinAdams) [reply src/utils/validation.js:198]: @danieljordan-caci  Done

File: src/components/form/OfficeAccountRequestFields/index.jsx [hunk 8]
```diff
@@ -0,0 +1,102 @@
+import React from 'react';
+import { func } from 'prop-types';
+import { Fieldset } from '@trussworks/react-uswds';
+
+import TextField from 'components/form/fields/TextField/TextField';
+import { CheckboxField, DutyLocationInput } from 'components/form/fields';
+import { searchTransportationOfficesOpen } from 'services/ghcApi';
+
+export const OfficeAccountRequestFields = ({ render }) => {
```
[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/components/form/OfficeAccountRequestFields/index.jsx:9]: @danieljordan-caci  Done

File: src/components/MilMoveHeader/LoggedOutUserInfo.jsx [hunk 9]
```diff
@@ -4,17 +4,36 @@ import { Button } from '@trussworks/react-uswds';
...
+        <div className={styles.userInfo}>
+          <ul className="usa-nav__primary">
+            <li className="usa-nav__primary-item">
+              <Button
+                unstyled
+                aria-label="Request Account"
+                onClick={handleRequestAccount}
+                data-testid="requestaccount"
```
[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/components/MilMoveHeader/LoggedOutUserInfo.jsx:18]: @danieljordan-caci  Done

File: src/utils/validation.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/utils/validation.js:192]: @danieljordan-caci  cool got it. I made the change have not merged any changes into the int branch yet.

File: src/pages/Office/RequestAccount/RequestAccount.jsx [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+      lastName: values.officeAccountRequestLastName,
+      telephone: values.officeAccountRequestTelephone,
+      transportationOfficeId: values.officeAccountTransportationOffice.id,
+      roles: requestedRoles,
+    };
+
+    return createOfficeAccountRequest({ body })
+      .then(() => {
+        navigate(generalRoutes.SIGN_IN_PATH);
+      })
```
[the remark this answers is not in the document]
Comment (TevinAdams) [reply src/pages/Office/RequestAccount/RequestAccount.jsx:78]: @danieljordan-caci added flash message

**Comment by TevinAdams** [comment]
> Looking at the blurb you have on the page - have you had this reviewed by Danny? Might want to make sure he's cool with what's written. Ask me how I know haha
> 
> ```
> To request an office acount, please fill out the form below with all required fields. \
> Once submitted, a MilMove admin will review your account. 
> If approved, you will be notified and allowed to log in.
> 
> NOTE: When filling out your EDIPI or unique identifier, you MUST provide one or the other. 
> If using CAC, fill out the EDIPI field. If using PIV or ECA certs, please fill out the unique identifier.
> ```

Changed it to the blurb you suggested. No longer using the edipi term.

**Comment by TevinAdams** [comment]
> You have a TON of commits that need to be looked at. A lot of these should already be in `integrationTesting` - do you happen to have some `main` commits in here?

This branch does have the branch for the open transportation office endpoint merged into it.

**Comment by danieljordan-caci** [comment]
> > You have a TON of commits that need to be looked at. A lot of these should already be in `integrationTesting` - do you happen to have some `main` commits in here?
> 
> This branch does have the branch for the open transportation office endpoint merged into it.

It seems like a lot more than that. Any input on this @cameroncaci or @deandreJones?

**Comment by TevinAdams** [comment]
> > > You have a TON of commits that need to be looked at. A lot of these should already be in `integrationTesting` - do you happen to have some `main` commits in here?
> > 
> > 
> > This branch does have the branch for the open transportation office endpoint merged into it.
> 
> It seems like a lot more than that. Any input on this @cameroncaci or @deandreJones?

I also had to merge in Cameron's branch with the new endpoint to submit the form.

**Comment by danieljordan-caci** [comment]
> > > > You have a TON of commits that need to be looked at. A lot of these should already be in `integrationTesting` - do you happen to have some `main` commits in here?
> > > 
> > > 
> > > This branch does have the branch for the open transportation office endpoint merged into it.
> > 
> > 
> > It seems like a lot more than that. Any input on this @cameroncaci or @deandreJones?
> 
> I also had to merge in Cameron's branch with the new endpoint to submit the form.

Yeah but all of that should already be in `integrationTesting` already so it shouldn't be appearing here

**Comment by traskowskycaci** [comment]
> Yeah but all of that should already be in `integrationTesting` already so it shouldn't be appearing here

Pulling in the latest int might make this easier to review, right? Should bring down the # of commits

**Comment by TevinAdams** [comment]
> > > > > You have a TON of commits that need to be looked at. A lot of these should already be in `integrationTesting` - do you happen to have some `main` commits in here?
> > > > 
> > > > 
> > > > This branch does have the branch for the open transportation office endpoint merged into it.
> > > 
> > > 
> > > It seems like a lot more than that. Any input on this @cameroncaci or @deandreJones?
> > 
> > 
> > I also had to merge in Cameron's branch with the new endpoint to submit the form.
> 
> Yeah but all of that should already be in `integrationTesting` already so it shouldn't be appearing here

I did merge it into my branch before it was in integration.

**Review by cameroncaci — CHANGES_REQUESTED** [review]
@TevinAdams 
Just make a fresh branch off integrationTesting -> cherry-pick your commits explicitly meant for 18830 into that new branch -> push and new PR. The git history here is pretty polluted, doing this would be the easiest way to start fresh. No need for any branch mergres since this one's upstream is already in integrationTesting

**Pull request closed without merging**